sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/JavaTypeInference.scala

```diff
@@ -212,82 +219,90 @@ object JavaTypeInference {
 
       case c if c == classOf[java.lang.String] =>
-        Invoke(path, "toString", ObjectType(classOf[String]))
+        createDeserializerForString(path, returnNullable = true)
 
       case c if c == classOf[java.math.BigDecimal] =>
-        Invoke(path, "toJavaBigDecimal", ObjectType(classOf[java.math.BigDecimal]))
+        createDeserializerForJavaBigDecimal(path, returnNullable = true)
+
+      case c if c == classOf[java.math.BigInteger] =>
+        createDeserializerForJavaBigInteger(path, returnNullable = true)
 
       case c if c.isArray =>
         val elementType = c.getComponentType
-        val primitiveMethod = elementType match {
-          case c if c == java.lang.Boolean.TYPE => Some("toBooleanArray")
-          case c if c == java.lang.Byte.TYPE => Some("toByteArray")
-          case c if c == java.lang.Short.TYPE => Some("toShortArray")
-          case c if c == java.lang.Integer.TYPE => Some("toIntArray")
-          case c if c == java.lang.Long.TYPE => Some("toLongArray")
-          case c if c == java.lang.Float.TYPE => Some("toFloatArray")
-          case c if c == java.lang.Double.TYPE => Some("toDoubleArray")
-          case _ => None
-        }
-
-        primitiveMethod.map { method =>
-          Invoke(path, method, ObjectType(c))
-        }.getOrElse {
-          Invoke(
-            MapObjects(
-              p => deserializerFor(typeToken.getComponentType, p),
-              path,
-              inferDataType(elementType)._1),
-            "array",
-            ObjectType(c))
+        val newTypePath = s"""- array element class: "${elementType.getCanonicalName}"""" +:
+          walkedTypePath
+        val (dataType, elementNullable) = inferDataType(elementType)
+
+        if (elementNullable) {
+          val mapFunction: Expression => Expression = element => {
+            // upcast the array element to the data type the encoder expected.
+            deserializerForWithNullSafetyAndUpcast(
+              element,
+              dataType,
+              nullable = elementNullable,
+              newTypePath,
+              (casted, typePath) => deserializerFor(typeToken.getComponentType, casted, typePath))
+          }
+          val arrayData = MapObjects(mapFunction, path, dataType)
+          Invoke(arrayData, "array", ObjectType(c), returnNullable = false)
+        } else {
+          val primitiveMethod = elementType match {
+            case c if c == java.lang.Integer.TYPE => "toIntArray"
+            case c if c == java.lang.Long.TYPE => "toLongArray"
+            case c if c == java.lang.Double.TYPE => "toDoubleArray"
+            case c if c == java.lang.Float.TYPE => "toFloatArray"
+            case c if c == java.lang.Short.TYPE => "toShortArray"
+            case c if c == java.lang.Byte.TYPE => "toByteArray"
+            case c if c == java.lang.Boolean.TYPE => "toBooleanArray"
+            case other => throw new IllegalStateException("expect primitive array element type " +
```

Comment:
Why we don't allow non-nullable & non-primitive case?

Reply:
Ah my bad. It was for sync between ScalaReflection and JavaTypeInference and I realized it is not exactly same. I'll roll back the change.

Reply:
Thanks!

Reply:
btw, we don't have any test for this code path? (it seems the latest Jenkins test run passed though...)

Reply:
I guess it is very hard to cover all cases, especially in Java test we rely on defining actual bean class to verify. It would be nice to cover more cases if we could define (column type, field type) matrix and create classes. Maybe another JIRA issue for this?

Reply:
WDYT, @cloud-fan ?

Reply:
@maropu 
Sorry to make a confusion. I wasn't wrong so the change was unnecessary, so I'm rolling back the change. To answer the origin comment, non-nullable & non-primitive case => not possible according to the implementation of `inferDataType`. `nullable = true` for only primitive case.

Reply:
> non-nullable & non-primitive case => not possible

I agree with this, and I think the original code style is better, i.e.
```
val primitiveMethod = ...
val primitiveMethod.map {
  ...
}.getOrElse {
  ...
}
```

We can change the scala side to follow this.

Reply:
OK. Would we keep `mapFunction` to deal with upcast? I guess `primitiveMethod.map.getOrElse` gets rid of necessary of `if (elementNullable)`, but just confirm again whether we would want to go back to origin code which upcast was not there, or just take previous code style.

Reply:
we definitely need to do upcast, just take the code style.

Reply:
Thanks for clarification! Just addressed.